framework/src/reporters/ReporterData.C

```diff
@@ -15,138 +15,157 @@ ReporterData::ReporterData(MooseApp & moose_app) : _app(moose_app) {}
-      oss << "\n    " << name;
-    mooseError(oss.str());
-  }
+  if (missing.size())
+    mooseError("The following Reporter(s) were not declared:\n\n", missing);
 }
 
 RestartableDataValue &
 ReporterData::getRestartableDataHelper(std::unique_ptr<RestartableDataValue> data_ptr,
                                        bool declare) const
 {
   // get the name to avoid problems with arbitrary function argument evaluation
-  std::string name = data_ptr->name();
+  const auto name = data_ptr->name();
   return _app.registerRestartableData(name, std::move(data_ptr), 0, !declare);
 }
 
 bool
 ReporterData::hasReporterWithMode(const std::string & obj_name, const ReporterMode & mode) const
 {
-  for (auto & context_ptr : _context_ptrs)
-  {
-    const ReporterName & rname = context_ptr->name();
-    if (rname.getObjectName() == obj_name && context_ptr->getProducerModeEnum() == mode)
+  for (const auto & name_context_pair : _context_ptrs)
+    if (name_context_pair.first.getObjectName() == obj_name &&
+        name_context_pair.second->getProducerModeEnum() == mode)
       return true;
-  }
   return false;
 }
 
-void
-ReporterData::transfer(const ReporterName & from_name,
-                       const ReporterName & to_name,
-                       ReporterData & to_data,
-                       unsigned int to_index) const
+const ReporterProducerEnum &
+ReporterData::getReporterMode(const ReporterName & reporter_name) const
 {
-  const ReporterContextBase * ptr = getReporterContextBase(from_name);
-  if (ptr == nullptr)
-    mooseError("Unable to locate Reporter with name:", from_name);
-  ptr->transfer(to_data, to_name, to_index);
+  return getReporterContextBase(reporter_name).getProducerModeEnum();
 }
 
-void
-ReporterData::resize(const ReporterName & name, dof_id_type n)
+bool
+ReporterData::hasReporterState(const ReporterName & reporter_name) const
 {
-  auto func = [name](const std::unique_ptr<ReporterContextBase> & ptr) {
-    return ptr->name() == name;
-  };
-  auto ptr = std::find_if(_context_ptrs.begin(), _context_ptrs.end(), func);
-  if (ptr == _context_ptrs.end())
-    mooseError("Unable to locate Reporter with name:", name);
-  (*ptr)->resize(n);
+  return _states.count(reporter_name);
 }
 
-void
-ReporterData::addConsumerMode(ReporterMode mode, const std::string & object_name)
+std::string
+ReporterData::getReporterInfo(const ReporterStateBase & state, const ReporterContextBase * context)
+{
+  std::stringstream oss;
+
+  const auto & name = state.getReporterName();
+
+  if (name.isPostprocessor())
+    oss << "Postprocessor \"" << name.getObjectName() << "\":\n";
+  else
+  {
+    oss << name.specialTypeToName() << " \"" << name.getCombinedName() << "\":\n  Type:\n    "
+        << state.valueType() << "\n";
+  }
+  oss << "  Producer:\n    ";
+  if (context)
+  {
+    oss << context->getProducer().type() << " \"" << context->getProducer().name() << "\"";
+    oss << "\n  Context type:\n    " << context->contextType();
```

Comment:
All that virtual stuff is all for this one method. Can we just used the C++ type stuff for this one piece of information?